File: clients/java/src/main/java/spb_ai_champ/model/Specialty.java
package spb_ai_champ.model;

import spb_ai_champ.util.StreamUtil;

/**
 * Player's specialty
 */
public enum Specialty {
    /**
     * Logistics. Increased travel distance
     */
    LOGISTICS(0),
    /**
     * Production. Increased work speed
     */
    PRODUCTION(1),
    /**
     * Combat. Increased damage
     */
    COMBAT(2);

    public int tag;

    Specialty(int tag) {
        this.tag = tag;
    }

    /**
     * Read Specialty from input stream
     */
    public static Specialty readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
        case 0:
            return LOGISTICS;
        case 1:
            return PRODUCTION;
        case 2:
            return COMBAT;
        default:
            throw new java.io.IOException("Unexpected tag value");
        }
    }
}

File: clients/kotlin/src/main/kotlin/spb_ai_champ/model/Specialty.kt
package spb_ai_champ.model

import spb_ai_champ.util.StreamUtil

/**
 * Player's specialty
 */
enum class Specialty private constructor(val tag: Int) {
    /**
     * Logistics. Increased travel distance
     */
    LOGISTICS(0),
    /**
     * Production. Increased work speed
     */
    PRODUCTION(1),
    /**
     * Combat. Increased damage
     */
    COMBAT(2);

    companion object {
        /**
         * Read Specialty from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Specialty {
            return when (StreamUtil.readInt(stream)) {
            LOGISTICS.tag -> LOGISTICS
            PRODUCTION.tag -> PRODUCTION
            COMBAT.tag -> COMBAT
            else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }
}